### PR TITLE
sticky header for mobile

### LINKED ISSUE #2705: Make header sticky on scroll on mobile
The OFN Header (pic below) should be sticky on mobile and tablet, same as it is on desktop

image

This issue should be considered alongside #2713 in case they need to interact

### PR DESCRIPTION
**Pull request by ja9-look**
#### What? Why?

Closes #2705

The OFN Header should be sticky on mobile and tablet so that it's more convenient for users to access the home page/ collapsible menu.

#### What should we test?
- If the header maintains stickiness
- If the collapsible menu/ shopping cart still works accordingly
- No content should be covered by the sticky headers
- Sticky header should be pushing content down

(Below added by @kristinalim:)

Particular scenarios that need to be checked in both desktop and mobile:

- Scrolling down while the alert in the home page is open
- Closing the alert in the home page while scrolled down
- Opening and closing the off-canvas menu while the home page alert is open
- Opening and closing the off-canvas menu while the home page alert is closed

#### Release notes

(Below added by @kristinalim:)

- Make non-admin header stick to the top of the screen in both desktop and mobile devices.

Changelog Category: Fixed

#### How is this related to the Spree upgrade?
N/a

#### Discourse thread
N/a

#### Dependencies
N/a

#### Documentation updates
N/a

### CODE REVIEW

**Comment by ja9-look** [comment]
@sauloperez @luisramos0 @kristinalim  Thank you all for reviewing! I just added a small change to the h2 to fix the spacing by increasing the margin top.

**Comment by sauloperez** [comment]
Cool! Would you mind sharing a screenshot of the outcome?

**Comment by ja9-look** [comment]
Screenshot 2019-05-05 at 11 26 26

**Comment by CathyPearce** [comment]
Test failed:  Shop front heading is being covered by the sticky header.
Production: 
image
Staging:
image

**Comment by ja9-look** [comment]
Hi @luisramos0 - yes I believe it is ready for re-test.

**Comment by luisramos0** [comment]
ok, thanks @ja9-look 
re-review and then re-test :-)

**Comment by kristinalim** [comment]
Searching the code:

```
$ git grep -i main app/assets/stylesheets/ app/views/layouts/ | grep -i role
app/assets/stylesheets/darkswarm/header.css.scss:29:body > section[role='main'] {
app/assets/stylesheets/darkswarm/ui.css.scss:128:  [role="main"] {
app/views/layouts/darkswarm.html.haml:48:        %section{ role: "main" }
app/views/layouts/registration.html.haml:32:        %section{ role: "main" }
```

**Comment by Matt-Yorkley** [comment]
I've just had a play with this, and I think it needs more thought. A sticky header clashes with the sticky banner on the homepage, and it's not easy to make them play nicely together...

**Comment by kristinalim** [comment]
@Matt-Yorkley I believe I managed to get it working. Moved the margin to the outer wrapper so we don't have to apply it to multiple page-specific elements. Do take a look when you have a chance.

@ja9-look Thank you very much for the work you did here! :slightly_smiling_face: I went ahead and made some changes. Hopefully, all is good now and we can merge soon.

**Comment by luisramos0** [comment]
@kristinalim 👏 👏

**Comment by Matt-Yorkley** [comment]
And it doesn't conflict with the registration banner on the homepage?

**Comment by Matt-Yorkley** [comment]
There's a slight issue in that when the hamburger menu is opened, the navigation switches from fixed to  no fixed, and jumps back to the top of the page...

**Comment by kristinalim** [comment]
@Matt-Yorkley Agh, you are right. I'm moving this back to In Dev.

**Comment by kristinalim** [comment]
Moving this back to Code Review. I've also added some specific scenarios to test in the issue description.

Here's a recording of the behaviour:

20190724131200-sticky_header_desktop_mobile

File: app/assets/javascripts/darkswarm/layout/offcanvas.js.coffee [hunk 1]
```diff
@@ -0,0 +1,6 @@
+$ ->
+  menu = $(".left-off-canvas-menu")
+  setOffcanvasMenuHeight = ->
+    menu.height($(window).height())
+  $(window).on("resize", setOffcanvasMenuHeight)
+  setOffcanvasMenuHeight()
```
[the remark this answers is not in the document]
Comment (luisramos0) [reply app/assets/javascripts/darkswarm/layout/offcanvas.js.coffee]: maybe as a controller in the _offcanvas_menu.html.haml ?

[the remark this answers is not in the document]
Comment (kristinalim) [reply app/assets/javascripts/darkswarm/layout/offcanvas.js.coffee]: Thanks for the suggestion, @luisramos0. The change is now here: https://github.com/openfoodfoundation/openfoodnetwork/pull/3786/commits/def2693afc5584eba8639cdb9e76855f998d88c2

File: app/assets/javascripts/darkswarm/controllers/offcanvas_controller.js.coffee [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Darkswarm.controller "OffcanvasCtrl", ($scope, $window) ->
+  $scope.menu = $(".left-off-canvas-menu")
+
+  $scope.setOffcanvasMenuHeight = ->
+    $scope.menu.height($(window).height())
+
+  $scope.bind = ->
+    $(window).on("resize", $scope.setOffcanvasMenuHeight)
```
Comment (luisramos0) [inline app/assets/javascripts/darkswarm/controllers/offcanvas_controller.js.coffee]: looking at this now I remembered this: https://www.tutorialsteacher.com/angularjs/angularjs-window-service
They say it's better for testing. Currently we have 4 cases of $(window) and a few more of $window

and the same for $(".left-off-canvas-menu") where we have a mix of $('') and angular.element

It's not very important but maybe we can define a best practice and do it one way or another.

Comment (kristinalim) [reply app/assets/javascripts/darkswarm/controllers/offcanvas_controller.js.coffee]: @luisramos0 Ah, yes about `angular.element`, I should have done that.

About `$window`, I might be wrong but, from what I understand, it is not a full-featured equivalent of `window` but only has properties that we would normally want to stub in tests.

I initially used `$window.height`, but it turns out is not supported (`undefined`). I forgot to remove `$window` afterwards. This removes it: https://github.com/openfoodfoundation/openfoodnetwork/pull/3786/commits/dae23d3c3ba970b47c4ded0d8a3c2a739b051b7f

Thanks, @luisramos0!

Comment (luisramos0) [reply app/assets/javascripts/darkswarm/controllers/offcanvas_controller.js.coffee]: ah  cool, I didnt notice $window was in the deps list :-) 👍

**Comment by daniellemoorhead** [comment]
Works good, notes here: https://docs.google.com/document/d/1trKDuRQSVmqOjZ1wxSnbpknqc1TCfrmK6xhBYd5Wd3I/edit?usp=sharing

**Comment by luisramos0** [comment]
the testing team is growing! 💪

**Pull request merged**